whois-api/src/main/java/net/ripe/db/whois/api/rdap/RdapRelationService.java

```diff
@@ -0,0 +1,202 @@
+    }
+
+    private static List<IpEntry> findSiblingsAndExact(final IpTree ipTree, final IpInterval parentResource) {
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(parentResource);
+        if (parentList.isEmpty()){
+            return ipTree.findExact(parentResource);
+        }
+        return ipTree.findFirstMoreSpecific(IpInterval.parse(parentList.getFirst().getKey().toString()));
+    }
+
+    private IpEntry searchFirstLessSpecific(final IpTree ipTree, final IpInterval reverseIp){
+        final List<IpEntry> parentList = ipTree.findFirstLessSpecific(reverseIp);
+        if (parentList.isEmpty() || !resourceExist(parentList.getFirst())){
+            throw new RdapException("404 Not Found", "No up level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+        }
+        return parentList.getFirst();
+    }
+
+    private List<IpEntry> searchTopLevel(final IpTree ipTree, final IpInterval reverseIp) {
+        for (final Object parentEntry : ipTree.findAllLessSpecific(reverseIp)) {
+            final IpEntry ipEntry = (IpEntry) parentEntry;
+            if (resourceExist(ipEntry)){
+                return List.of(ipEntry);
+            }
+        }
+        throw new RdapException("404 Not Found", "No top level object has been found for " + reverseIp.toString(), HttpStatus.NOT_FOUND_404);
+    }
+
+    private boolean resourceExist(final IpEntry firstLessSpecific){
+        final RpslObject rpslObject = getResourceByKey(firstLessSpecific.getKey().toString());
+        if (rpslObject == null || isAdministrativeResource(rpslObject)) {
+            LOGGER.error("INET(6)NUM {} does not exist in RIPE Database ", firstLessSpecific.getKey().toString());
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean isAdministrativeResource(final RpslObject rpslObject) {
+        final CIString statusAttributeValue = rpslObject.getValueForAttribute(AttributeType.STATUS);
+        return (rpslObject.getType().equals(ObjectType.INETNUM) && InetnumStatus.getStatusFor(statusAttributeValue).equals(ALLOCATED_UNSPECIFIED))
+                || (rpslObject.getType().equals(ObjectType.INET6NUM) && Inet6numStatus.getStatusFor(statusAttributeValue).equals(ALLOCATED_BY_RIR));
+    }
+
+
+    @Nullable
+    private RpslObject getResourceByKey(final String key){
+        final RpslObject rpslObject = rpslObjectDao.getByKeyOrNull(ObjectType.INET6NUM, key);
+        if (rpslObject == null){
+            return rpslObjectDao.getByKeyOrNull(ObjectType.INETNUM, key);
+        }
+        return rpslObject;
+    }
+
+    private IpTree getIpTree(final IpInterval reverseIp) {
+        if (reverseIp instanceof Ipv4Resource) {
+            return ip4Tree;
+        } else if (reverseIp instanceof Ipv6Resource) {
+            return ip6Tree;
```

Comment:
No need for else if ? use ternary operator. There are only two impelmentations of IpInterval

Reply:
Done